File: Recursion/imp/SubsetSum.cpp
#include <bits/stdc++.h>

using namespace std;

void SubsetSum(int i , int n , vector<int> &vec , int arr[] , int s){
    if (i>=n){
        // cout<<s;
        vec.push_back(s);
        return ;
    }

    SubsetSum(i+1,n,vec,arr,s);
    SubsetSum(i+1,n,vec,arr,s+arr[i]);

}

int main(){
    int n;
    cin >> n;

    int arr[n];

    for (int i = 0 ; i<n ; i++){
        cin >> arr[i];
    }


    vector<int> vec;
    SubsetSum(0,n,vec,arr,0);

    sort(vec.begin() , vec.end());
    for (auto it:vec){
        cout<<it<<endl;
    }

    return 0;
}

File: DP/subsetSum.java
public class subsetSum {

    public static boolean f(int n, int [] arr, int i){
        if (n == 0) return true;
        if (i == arr.length) return false;

        boolean take = f(n - arr[i] , arr , i+1);
        boolean nottake = f(n  , arr , i+1);

        return take || nottake;
    }
    public static void main(String [] arfs){
        int n = 9;
        int [] arr = {0,7,8};
        
        System.out.println(f(n,arr,0));

    }
}
